Signal/src/util/OWSBackupExport.m

```diff
@@ -0,0 +1,601 @@
+{
+    OWSAssert(self.exportDirPath.length > 0);
+
+    // TODO: Encrypt the file using self.delegate.backupKey;
+
+    NSString *dstFilePath = [self.exportDirPath stringByAppendingPathComponent:[NSUUID UUID].UUIDString];
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSError *error;
+    BOOL success = [fileManager copyItemAtPath:srcFilePath toPath:dstFilePath error:&error];
+    if (!success || error) {
+        OWSProdLogAndFail(@"%@ error writing encrypted file: %@", self.logTag, error);
+        return nil;
+    }
+    return dstFilePath;
+}
+
+@end
+
+#pragma mark -
+
+@interface OWSBackupExport () <SSZipArchiveDelegate>
+
+@property (nonatomic, weak) id<OWSBackupExportDelegate> delegate;
+
+@property (nonatomic, nullable) YapDatabaseConnection *srcDBConnection;
+
+@property (nonatomic, nullable) YapDatabaseConnection *dstDBConnection;
+
+// Indicates that the backup succeeded, failed or was cancelled.
+@property (atomic) BOOL isComplete;
+
+@property (nonatomic, nullable) OWSBackupStorage *backupStorage;
+
+@property (nonatomic, nullable) NSData *databaseSalt;
+
+@property (nonatomic, nullable) OWSBackgroundTask *backgroundTask;
+
+@property (nonatomic) NSMutableArray<NSString *> *databaseFilePaths;
+// A map of "record name"-to-"file name".
+@property (nonatomic) NSMutableDictionary<NSString *, NSString *> *databaseRecordMap;
+
+// A map of "attachment id"-to-"local file path".
+@property (nonatomic) NSMutableDictionary<NSString *, NSString *> *attachmentFilePathMap;
+// A map of "record name"-to-"file relative path".
+@property (nonatomic) NSMutableDictionary<NSString *, NSString *> *attachmentRecordMap;
+
+@property (nonatomic, nullable) NSString *manifestFilePath;
+@property (nonatomic, nullable) NSString *manifestRecordName;
+
+@property (nonatomic) NSString *exportDirPath;
+
+@end
+
+#pragma mark -
+
+@implementation OWSBackupExport
+
+- (instancetype)initWithDelegate:(id<OWSBackupExportDelegate>)delegate
+                  primaryStorage:(OWSPrimaryStorage *)primaryStorage
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(primaryStorage);
+    OWSAssert([OWSStorage isStorageReady]);
+
+    self.delegate = delegate;
+    _srcDBConnection = primaryStorage.newDatabaseConnection;
+
+    return self;
+}
+
+- (void)dealloc
+{
+    // Surface memory leaks by logging the deallocation.
+    DDLogVerbose(@"Dealloc: %@", self.class);
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+
+    if (self.exportDirPath) {
+        [OWSFileSystem deleteFileIfExists:self.exportDirPath];
+    }
+}
+
+- (void)startAsync
+{
+    OWSAssertIsOnMainThread();
+
+    DDLogInfo(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    self.backgroundTask = [OWSBackgroundTask backgroundTaskWithLabelStr:__PRETTY_FUNCTION__];
+
+    __weak OWSBackupExport *weakSelf = self;
+    [OWSBackupAPI checkCloudKitAccessWithCompletion:^(BOOL hasAccess) {
```

Comment:
Backup export and import are highly async processes.  Here we check for iCloud access.